### PR TITLE
Add preliminary support to XD60 Rev3 PCB

### PR DESCRIPTION
**Pull request by edulpn**
Add preliminary support to XD60 Rev3 PCB (Supports the 12 underglow LEDs, but split spacebar still not wired/implemented).

The previous PCB revision QMK works fine if flashed but the RGB 12 LED underglow will not work (only controls the 6 bottom LEDs). I'm creating this so people can start working on extra features (e.g. Split Spacebar support) separate from the previous revision firmware.

### CODE REVIEW

**Comment by mechmerlin** [comment]
I'd prefer to have this as a `rev3` directory in `keyboards/xd60`, and the current implementation put in a `rev1`.

File: keyboards/xd60_rev3/xd60_rev3.c [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+void led_set_kb(uint8_t usb_led) {
+	// put your keyboard LED indicator (ex: Caps Lock LED) toggling code here
+
+  if (usb_led & (1<<USB_LED_CAPS_LOCK)) {
+		xd60_caps_led_on();
+	} else {
+		xd60_caps_led_off();
+  }
+
+  // if (usb_led & (1<<USB_LED_NUM_LOCK)) {
```
Comment (mechmerlin) [inline keyboards/xd60_rev3/xd60_rev3.c]: You can remove this uncommented code if it's not being used.

Comment (edulpn) [reply keyboards/xd60_rev3/xd60_rev3.c]: Of course!

**Comment by edulpn** [comment]
> I'd prefer to have this as a `rev3` directory in `keyboards/xd60`, and the current implementation put in a `rev1`.

Understood, I can certainly do this... what will be the `make` command syntax, so I can update the README?

**Comment by mechmerlin** [comment]
> > I'd prefer to have this as a `rev3` directory in `keyboards/xd60`, and the current implementation put in a `rev1`.
> 
> Understood, I can certainly do this... what will be the `make` command syntax, so I can update the README? This should also be updated on the XD60 current implementation as well, right?

Correct, you would be editing the currently existing xd60 files. 

Current make syntax is 
```
make xd60:default
```

after the changes I have suggested, it would be one of the following. 

```
make xd60/rev3:default
make xd60/rev1:default
```

**Comment by edulpn** [comment]
Understood, I'll make the changes and update the PR. Thanks for your time!

File: keyboards/xd60_rev3/config.h [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef CONFIG_H
```
Comment (drashna) [inline keyboards/xd60_rev3/config.h]: If you could replace the include guard (the ifndef, define, and endif at the end) with just `#pragma once`? 

It's simpler, and less prone to user error, so if you could make this switch? 

```suggestion
#pragma once
```

File: keyboards/xd60_rev3/config.h [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+#define CONFIG_H
```
Comment (drashna) [inline keyboards/xd60_rev3/config.h]:
```suggestion
```

File: keyboards/xd60_rev3/config.h [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+#define LOCKING_SUPPORT_ENABLE
+/* Locking resynchronize hack */
+#define LOCKING_RESYNC_ENABLE
+
+/* key combination for magic key command */
+#define IS_COMMAND() ( \
+  keyboard_report->mods == (MOD_BIT(KC_LSHIFT) | MOD_BIT(KC_RSHIFT)) \
+)
+
+#endif
```
Comment (drashna) [inline keyboards/xd60_rev3/config.h]:
```suggestion
```

File: keyboards/xd60_rev3/keymaps/default/keymap.c [hunk 5]
```diff
@@ -0,0 +1,46 @@
+#include QMK_KEYBOARD_H
+#include "action_layer.h"
```
Comment (drashna) [inline keyboards/xd60_rev3/keymaps/default/keymap.c]: This isn't actually needed, and is included by default already. 

```suggestion
```

File: keyboards/xd60_rev3/keymaps/default/keymap.c [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+
+  // 0: Base Layer
+  LAYOUT_all(
+      KC_ESC,  KC_1,    KC_2,    KC_3,    KC_4,   KC_5,   KC_6,   KC_7,   KC_8,   KC_9,    KC_0,    KC_MINS,  KC_EQL,  KC_BSPC,  KC_GRV,    \
+      KC_TAB,  KC_Q,    KC_W,    KC_E,    KC_R,   KC_T,   KC_Y,   KC_U,   KC_I,   KC_O,    KC_P,    KC_LBRC,  KC_RBRC,           KC_BSLS,   \
+      KC_CAPS, KC_A,    KC_S,    KC_D,    KC_F,   KC_G,   KC_H,   KC_J,   KC_K,   KC_L,    KC_SCLN, KC_QUOT,  KC_NO,             KC_ENT,    \
+      KC_LSFT, KC_NO,   KC_Z,    KC_X,    KC_C,   KC_V,   KC_B,   KC_N,   KC_M,   KC_COMM, KC_DOT,  KC_SLSH,  KC_RSFT, KC_UP,    KC_DEL,      \
+      KC_LCTL, KC_LGUI, KC_LALT,                          KC_SPC,                          KC_RGUI, F(0),     KC_LEFT, KC_DOWN,  KC_RIGHT),
```
Comment (drashna) [inline keyboards/xd60_rev3/keymaps/default/keymap.c]: Could you please remove the function, as this is deprecated, and there is a keycode to handle momentary layers.

```suggestion
      KC_LCTL, KC_LGUI, KC_LALT,                          KC_SPC,                          KC_RGUI, MO(1),     KC_LEFT, KC_DOWN,  KC_RIGHT),
```

File: keyboards/xd60_rev3/keymaps/default/keymap.c [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+  // 1: Function Layer
+  LAYOUT_all(
+      RESET,   KC_F1,   KC_F2,   KC_F3,   KC_F4,  KC_F5,  KC_F6,  KC_F7,  KC_F8,  KC_F9,   KC_F10,  KC_F11,   KC_F12,  KC_F13,   KC_F14,    \
+      KC_NO,   KC_WH_U, KC_UP,   KC_WH_D, KC_BSPC,KC_HOME,KC_CALC,KC_NO,  KC_INS, KC_NO,   KC_PSCR, KC_SLCK,  KC_PAUS,           KC_DEL,    \
+      KC_NO,   KC_LEFT, KC_DOWN, KC_RIGHT,KC_DEL, KC_END, KC_PGDN,KC_NO,  KC_NO,  KC_NO,   KC_HOME, KC_PGUP,  KC_NO,             KC_ENT,    \
+      KC_LSFT, KC_NO,   KC_NO,   KC_APP,  BL_STEP,KC_NO,  KC_NO,  KC_VOLD,KC_VOLU,KC_MUTE, KC_END,  KC_PGDN,  KC_RSFT, KC_PGUP,  KC_INS,      \
+      KC_LCTL, KC_LGUI, KC_LALT,                          KC_SPC,                          KC_RGUI, F(0),     KC_HOME, KC_PGDOWN,KC_END),
```
Comment (drashna) [inline keyboards/xd60_rev3/keymaps/default/keymap.c]: Could you please remove the function, as this is deprecated, and there is a keycode to handle momentary layers.


```suggestion
      KC_LCTL, KC_LGUI, KC_LALT,                          KC_SPC,                          KC_RGUI, MO(1),     KC_HOME, KC_PGDOWN,KC_END),
```

File: keyboards/xd60_rev3/keymaps/default/keymap.c [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+};
+
+// Custom Actions
+const uint16_t PROGMEM fn_actions[] = {
```
Comment (drashna) [inline keyboards/xd60_rev3/keymaps/default/keymap.c]: Could you remove the `fn_actions` block?  All of it's features have been added as keycodes, and been deprecated in favor of `process_record_user`.

File: keyboards/xd60_rev3/keymaps/default/keymap.c [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    [0] = ACTION_LAYER_MOMENTARY(1),  // to Fn overlay
+};
+
+// Macros
+const macro_t *action_get_macro(keyrecord_t *record, uint8_t id, uint8_t opt) {
```
Comment (drashna) [inline keyboards/xd60_rev3/keymaps/default/keymap.c]: Could you remove the `action_get_macro` function?  It's been deprecated in favor of `process_record_user`. 

Additionally, it appears as if you're not actually using the macro, so it definitely could be removed safely.

File: keyboards/xd60_rev3/config.h (hunk 2 above)
Comment (edulpn) [reply keyboards/xd60_rev3/config.h]: Sure!

File: keyboards/xd60_rev3/keymaps/default/keymap.c (hunk 5 above)
Comment (edulpn) [reply keyboards/xd60_rev3/keymaps/default/keymap.c]: Sure!

File: keyboards/xd60_rev3/keymaps/default/keymap.c (hunk 6 above)
Comment (edulpn) [reply keyboards/xd60_rev3/keymaps/default/keymap.c]: Sure!

File: keyboards/xd60_rev3/keymaps/default/keymap.c (hunk 8 above)
Comment (edulpn) [reply keyboards/xd60_rev3/keymaps/default/keymap.c]: Will do

File: keyboards/xd60_rev3/keymaps/default/keymap.c (hunk 9 above)
Comment (edulpn) [reply keyboards/xd60_rev3/keymaps/default/keymap.c]: Will do

**Comment by edulpn** [comment]
I had not, but I'm planning to. The PR is getting in the way, right?

**Comment by drashna** [comment]
No, not in the way, at all.  However, it's been sitting for awhile, and just wanted to touch base since there had been no update in a couple of weeks.  :)

**Comment by edulpn** [comment]
Yeah, I haven't been able to get back to this PR, sorry.

**Comment by drashna** [comment]
No worries. I just wanted to make sure

**Comment by rodrigofeijao** [comment]
Are there anything else on this PR to be fixed? I'm ok to help with anything missing.

**Comment by ff4500** [comment]
@rodrigofeijao If you can shoehorn in split spacebar support, it would be appreciated. I think that may be the only additional thing that hasn't been addressed yet. @edulpn should be able to confirm, but it looks like just adding in the requested changes above.

**Comment by edulpn** [comment]
Yeah, the end of last year was a nightmare, I wasn't able to touch this PR. The changes requested were mostly outdated stuff I copied and pasted from the Rev2 firmware, but some stuff need to be moved around. I don't need help, per se, I'm just a bit busy... sorry guys :(

**Comment by rodrigofeijao** [comment]
@ff4500 I think that might be better to create another keyboard layout for the split spacebar, as it's only a layout change, am I right? If so, I'm totally happy to create that layout.

@edulpn I can totally understand the Christmas busy time as I had the same problem. But have some spare time now, so I've cloned your branch and did some changes to apply a few suggestions from @drashna, but unfortunately, I can't push those changes as I have no write permissions on that repo. 

And I think we should not concern about some suggestions @drashna did as they are related to the default layout, and to change that it might be better to be done on a different PR as it affects all revs, and this PR is only to add the support for REV3.

**Comment by ff4500** [comment]
@rodrigofeijao Creating the layout would be great, if you have the time. I don't know what it entails, since the original v2 layouts don't support it. Assuming that it's just a keymap/layout change to include the additional option(s) seen here?

htb1_qrwxh2rk1rksnhjq6ykdpxa0

**Comment by rodrigofeijao** [comment]
@ff4500 Creating a layout for each of those would be mental. But drop me a pm if you need to create a layout for your rev3, I can give a hand on that.

**Comment by ff4500** [comment]
> @ff4500 Creating a layout for each of those would be mental. But drop me a pm if you need to create a layout for your rev3, I can give a hand on that.

Completely understandable. Maybe a single split layout to help with the matrix for others?
Will drop you a PM shortly with what I'm personally looking for.

**Comment by drashna** [comment]
@edulpn Any update on this?

**Comment by rodrigofeijao** [comment]
@drashna It looks like this have been around for too long, should I do my own PR with the support for this?

**Comment by edulpn** [comment]
Hey guys, sorry for not being able to get back to this PR. @rodrigofeijao are you up to tackle this thing? If so, I'd prefer to pass up the baton, as I'm not really sure when I'll be able to get back to this in the future. Can I just close the pull request?

**Comment by drashna** [comment]
@edulpn Not a problem.  I'll go ahead and close this for you. 

@rodrigofeijao you should be able to pull the changes from either @edulpn's branch, or by just pulling the diff here and using `git apply` to apply the changes.
https://github.com/qmk/qmk_firmware/pull/4457.diff

**Comment by rodrigofeijao** [comment]
@drashna PR open at https://github.com/qmk/qmk_firmware/pull/4938

**Pull request closed without merging**